### PR TITLE
feat: Add shuffle function to Array

### PR DESCRIPTION
**Pull request by qazxcdswe123**
close: #488 

Not sure if an API like this is what you want, and I also think we should provide a default rand function for the user so I added the rand package and make the `rand` argument optional.

### CODE REVIEW

File: array/array.mbt [hunk 1]
```diff
@@ -79,3 +79,15 @@ pub fn Array::new_with_index[T](length : Int, value : (Int) -> T) -> Array[T] {
...
+
+pub fn Array::shuffle[T](arr : Array[T], ~rand : (Int, Int) -> Int) -> Unit {
+  let n = arr.length()
+  let mut i = n - 1
+  while i > 0 {
+    let j = rand(0, i + 1)
+    let tmp = arr[i]
+    arr[i] = arr[j]
+    arr[j] = tmp
+    i = i - 1
```
Comment (Frank-III) [inline array/array.mbt:91]: I think there is a `swap` method?

Comment (qazxcdswe123) [reply array/array.mbt:94]: Yes, thank you for the reminder

File: array/array.mbt [hunk 2]
```diff
@@ -79,3 +79,18 @@ pub fn Array::new_with_index[T](length : Int, value : (Int) -> T) -> Array[T] {
...
 }
+
+fn rand(upper : Int) -> Int {
+  let summoner = @random.RandomInt::new()
+  summoner.summon() % upper
+}
+
+pub fn Array::shuffle[T](arr : Array[T], ~rand : (Int) -> Int = rand) -> Unit {
+  let n = arr.length()
+  let mut i = n - 1
```
Comment (bobzhang) [inline array/array.mbt:90]: we tend to avoid mut if possible.
e.g,
```
for i = n - 1; i >0 ; i = i - 1 {}
```

File: array/moon.pkg.json [hunk 3]
```diff
@@ -2,6 +2,7 @@
   "import": [
     "moonbitlang/core/builtin",
     "moonbitlang/core/coverage",
-    "moonbitlang/core/assertion"
+    "moonbitlang/core/assertion",
+    "moonbitlang/core/random"
   ]
```
Comment (bobzhang) [inline array/moon.pkg.json:7]: we don't want array depend on random

File: array/array.mbti [hunk 4]
```diff
@@ -40,6 +40,7 @@ impl Array {
   blit_to[A](Self[A], Self[A], ~len : Int, ~src_offset : Int = .., ~dst_offset : Int = ..) -> Unit
   copy[T](Self[T]) -> Self[T]
   new_with_index[T](Int, (Int) -> T) -> Self[T]
+  shuffle[T](Self[T], ~rand : (Int) -> Int = ..) -> Unit
   sort[T : Compare + Eq](Self[T]) -> Unit
   sort_by[T](Self[T], (T, T) -> Int) -> Unit
```
Comment (bobzhang) [inline array/array.mbti:45]: let's provide both `shuffle` and `shuffle_in_place`

**Comment by bobzhang** [comment]
@qazxcdswe123 lgtm, I added two doc strings and a safety check

**Review by bobzhang — APPROVED** [review]

**Pull request merged**